perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/VosManagerBlImpl.java

```diff
@@ -902,14 +929,65 @@ public boolean usesEmbeddedGroupRegistrations(PerunSession sess, Vo vo) {
 
 	@Override
 	public void addMemberVo(PerunSession sess, Vo vo, Vo memberVo) throws RelationExistsException {
-		// todo - add necessary logic for adding new member vo
 		vosManagerImpl.addMemberVo(sess, vo, memberVo);
+
+		List<Member> memberVoMembers = perunBl.getMembersManagerBl().getMembers(sess, memberVo);
+		AttributeDefinition memberExpirationAttrDef;
+
+		try {
+			memberExpirationAttrDef = perunBl.getAttributesManagerBl().getAttributeDefinition(sess, AttributesManager.NS_MEMBER_ATTR_DEF + ":membershipExpiration");
+		} catch (AttributeNotExistsException e) {
+			throw new InternalErrorException(e);
+		}
+
+		for (Member member : memberVoMembers) {
+			try {
+
+				try {
+					// if user is member in both vos, update memberOrganizations list attribute, validate and reset expiration
+					Member existingMember = perunBl.getMembersManagerBl().getMemberByUserId(sess, vo, member.getUserId());
+					perunBl.getAttributesManagerBl().removeAttribute(sess, existingMember, memberExpirationAttrDef);
+					if (existingMember.getStatus().equals(Status.INVALID) || existingMember.getStatus().equals(Status.DISABLED)) {
+						perunBl.getMembersManagerBl().validateMemberAsync(sess, member);
+					}
+					Attribute attribute = perunBl.getAttributesManagerBl().getAttribute(sess, existingMember, A_MEMBER_DEF_MEMBER_ORGANIZATIONS);
+					ArrayList<String> currentValue = attribute.valueAsList();
+					currentValue.add(memberVo.getShortName());
+					attribute.setValue(currentValue);
+					perunBl.getAttributesManagerBl().setAttribute(sess, existingMember, attribute);
+
+				} catch (MemberNotExistsException e) {
+					// if user is member only in member vo, create member in parent vo and set up memberOrganizations list attribute
+					Member newMember = perunBl.getMembersManagerBl().createMember(sess, vo, perunBl.getUsersManagerBl().getUserByMember(sess, member));
```

Comment:
I just realized that creating members also sets expiration times for them according to the parent VO rules. Therefore, we should most likely unset these expiration times for them because they should follow expiration rules only when they are not valid in any member organization.